src/Features/Core/Portable/Completion/CompletionServiceWithProviders.ProviderSource.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    public abstract partial class CompletionServiceWithProviders
+    {
+        private class ProviderSource : IEqualityComparer<ImmutableHashSet<string>>
+        {
+            private readonly CompletionServiceWithProviders _service;
+            private IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>>? _importedProviders;
+
+            private readonly object _gate = new();
+            private readonly Dictionary<string, CompletionProvider?> _nameToProvider = new();
+            private readonly Dictionary<ImmutableHashSet<string>, Task<ImmutableArray<CompletionProvider>>> _rolesToProviders;
+
+            public ProviderSource(CompletionServiceWithProviders service)
+            {
+                _service = service;
+                _rolesToProviders = new(this);
+            }
+
+            public ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, bool waitUntilAvaialble = false)
+            {
+                Task<ImmutableArray<CompletionProvider>>? createProviderTask;
+
+                lock (_gate)
+                {
+                    if (!_rolesToProviders.TryGetValue(roles, out createProviderTask))
+                    {
+                        createProviderTask = GetAllProvidersAsync(roles);
+                        _rolesToProviders[roles] = createProviderTask;
+                    }
+                }
+
+                return createProviderTask.IsCompleted || waitUntilAvaialble
+                    ? createProviderTask.Result
+                    : ImmutableArray<CompletionProvider>.Empty;
+            }
+
+            public CompletionProvider? GetProviderByName(string name)
+            {
+                lock (_gate)
+                {
+                    _nameToProvider.TryGetValue(name, out var provider);
+                    return provider;
+                }
+            }
+
+            private async Task<ImmutableArray<CompletionProvider>> GetAllProvidersAsync(ImmutableHashSet<string> roles)
+            {
+                await Task.Yield();
```

Comment:
📝 This will continue on the current synchronization context. Did you mean to have `ConfigureAwait(false)`?

Reply:
Yes. Missing this completely defeated all the changes made here :P

Reply:
i'm also not sure how this helps.  won't this be called when the editor is askign us for completion items?  aren't they already calling into us on the BG?

Reply:
This is also called on UI thread by command handler handling keystrokes, to decide whether we should trigger completion. You can see some of problematic the callstack [here](https://prism.vsdata.io/failure/?eventType=UIDelay&failureHash=51577563-0f16-9e07-4ee8-760bd63cc6b0)

Reply:
ick.  should we pass along to VS that they should move this to the BG as well?

Reply:
looks like we need this to be on UI thread
https://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs,82